gapic/src/platform/windows/org/eclipse/swt/widgets/SwtUtil.java

```diff
@@ -15,11 +15,64 @@
  */
 package org.eclipse.swt.widgets;
 
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.internal.win32.OS;
+import org.eclipse.swt.internal.win32.SCROLLINFO;
+
 public class SwtUtil {
   private SwtUtil() {
   }
 
   public static void disableAutoHideScrollbars(@SuppressWarnings("unused") Scrollable widget) {
     // Do nothing.
   }
+
+  public static void syncTreeAndTableScroll(Tree tree, Table table) {
+    Exclusive exclusive = new Exclusive();
+    table.getVerticalBar().addListener(SWT.Selection, e ->
+      exclusive.runExclusive(() -> {
+        int pos = table.getVerticalBar().getSelection();
+        SCROLLINFO info = new SCROLLINFO ();
+        info.cbSize = SCROLLINFO.sizeof;
+        info.fMask = OS.SIF_POS;
+        info.nPos = pos;
+        OS.SetScrollInfo(tree.handle, OS.SB_VERT, info, true);
+        OS.SendMessage(tree.handle, OS.WM_VSCROLL, OS.SB_THUMBPOSITION | (pos << 16), 0);
+      }));
+
+    Runnable updateTableScroll = () -> {
+      exclusive.runExclusive(() -> table.setTopIndex(tree.getVerticalBar().getSelection()));
+    };
+    tree.getVerticalBar().addListener(SWT.Selection, e -> updateTableScroll.run());
+    tree.addListener(SWT.Expand, e -> table.getDisplay().asyncExec(updateTableScroll));
+    tree.addListener(SWT.Collapse, e -> table.getDisplay().asyncExec(updateTableScroll));
+
+    // Make sure the rows are the same height in the table as the tree.
+    int[] height = { tree.getItemHeight() };
+    table.addListener(SWT.Paint, event -> {
+      height[0] = tree.getItemHeight();
+      if (table.getItemHeight() != height[0]) {
+        table.setItemHeight(height[0]);
+        updateTableScroll.run();
+      }
+    });
+  }
+
+  private static class Exclusive {
+    private boolean ignore = false;
```

Comment:
Is this operating as a mutex or prevention against recursion or reentrancy? Can you add some description of what it does and why it's needed?

Reply:
Added a comment.